File: src/main/java/edu/kit/iti/formal/stvs/model/code/SyntaxErrorListener.java
package edu.kit.iti.formal.stvs.model.code;

import java.util.ArrayList;
import java.util.BitSet;
import java.util.List;

import org.antlr.v4.runtime.ANTLRErrorListener;
import org.antlr.v4.runtime.Parser;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.Recognizer;
import org.antlr.v4.runtime.atn.ATNConfigSet;
import org.antlr.v4.runtime.dfa.DFA;

/**
 * A listener for {@link SyntaxError}s in code.
 *
 * @author Lukas Fritsch
 */
public class SyntaxErrorListener implements ANTLRErrorListener {
  private List<SyntaxError> syntaxErrors = new ArrayList<>();

  @Override
  public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line,
      int charPositionInLine, String msg, RecognitionException exc) {
    syntaxErrors.add(new SyntaxError(line, charPositionInLine, msg));
  }

  public List<SyntaxError> getSyntaxErrors() {
    return syntaxErrors;
  }


  @Override
  public void reportAmbiguity(Parser recognizer, DFA dfa, int startIndex, int stopIndex,
      boolean exact, BitSet ambigAlts, ATNConfigSet configs) {

  }

  @Override
  public void reportAttemptingFullContext(Parser recognizer, DFA dfa, int startIndex, int stopIndex,
      BitSet conflictingAlts, ATNConfigSet configs) {

  }

  @Override
  public void reportContextSensitivity(Parser recognizer, DFA dfa, int startIndex, int stopIndex,
      int prediction, ATNConfigSet configs) {}
}

File: stvs/src/main/kotlin/edu/kit/iti/formal/stvs/model/code/SyntaxErrorListener.kt
package edu.kit.iti.formal.stvs.model.code

import org.antlr.v4.runtime.ANTLRErrorListener
import org.antlr.v4.runtime.Parser
import org.antlr.v4.runtime.RecognitionException
import org.antlr.v4.runtime.Recognizer
import org.antlr.v4.runtime.atn.ATNConfigSet
import org.antlr.v4.runtime.dfa.DFA
import java.util.*

/**
 * A listener for [SyntaxError]s in code.
 *
 * @author Lukas Fritsch
 */
class SyntaxErrorListener : ANTLRErrorListener {
    val syntaxErrors: MutableList<SyntaxError> = ArrayList()

    override fun syntaxError(
        recognizer: Recognizer<*, *>?, offendingSymbol: Any, line: Int,
        charPositionInLine: Int, msg: String, exc: RecognitionException
    ) {
        syntaxErrors.add(SyntaxError(line, charPositionInLine, msg))
    }

    override fun reportAmbiguity(
        recognizer: Parser, dfa: DFA, startIndex: Int, stopIndex: Int,
        exact: Boolean, ambigAlts: BitSet, configs: ATNConfigSet
    ) {
    }

    override fun reportAttemptingFullContext(
        recognizer: Parser, dfa: DFA, startIndex: Int, stopIndex: Int,
        conflictingAlts: BitSet, configs: ATNConfigSet
    ) {
    }

    override fun reportContextSensitivity(
        recognizer: Parser, dfa: DFA, startIndex: Int, stopIndex: Int,
        prediction: Int, configs: ATNConfigSet
    ) {
    }
}
